# Multiple Linear Regression in Statsmodels - Lab

## Introduction
In this lab, you'll practice fitting a multiple linear regression model on the Boston Housing dataset!

## Objectives
You will be able to:
* Determine if it is necessary to perform normalization/standardization for a specific model or set of data
* Use standardization/normalization on features of a dataset
* Identify if it is necessary to perform log transformations on a set of features
* Perform log transformations on different features of a dataset
* Use statsmodels to fit a multiple linear regression model
* Evaluate a linear regression model by using statistical performance metrics pertaining to overall model and specific parameters


## The Boston Housing Data

We pre-processed the Boston Housing data again. This time, however, we did things slightly different:
- We dropped `'ZN'` and `'NOX'` completely 
- We categorized `'RAD'` in 3 bins and `'TAX'` in 4 bins
- We transformed `'RAD'` and `'TAX'` to dummy variables and dropped the first variable to eliminate multicollinearity
- We used min-max-scaling on `'B'`, `'CRIM'`, and `'DIS'` (and log transformed all of them first, except `'B'`)
- We used standardization on `'AGE'`, `'INDUS'`, `'LSTAT'`, and `'PTRATIO'` (and log transformed all of them first, except for `'AGE'`) 

In [48]:
import pandas as pd
import numpy as np
from sklearn.datasets import load_boston
boston = load_boston()

boston_features = pd.DataFrame(boston.data, columns = boston.feature_names)
boston_features = boston_features.drop(['NOX', 'ZN'],axis=1)

# First, create bins for based on the values observed. 3 values will result in 2 bins
bins = [0, 6, 24]
bins_rad = pd.cut(boston_features['RAD'], bins)
bins_rad = bins_rad.cat.as_unordered()

# First, create bins for based on the values observed. 4 values will result in 3 bins
bins = [0, 270, 360, 712]
bins_tax = pd.cut(boston_features['TAX'], bins)
bins_tax = bins_tax.cat.as_unordered()

tax_dummy = pd.get_dummies(bins_tax, prefix='TAX', drop_first=True)
rad_dummy = pd.get_dummies(bins_rad, prefix='RAD', drop_first=True)
boston_features = boston_features.drop(['RAD','TAX'], axis=1)
boston_features = pd.concat([boston_features, rad_dummy, tax_dummy], axis=1)

In [49]:
age = boston_features['AGE']
b = boston_features['B']
logcrim = np.log(boston_features['CRIM'])
logdis = np.log(boston_features['DIS'])
logindus = np.log(boston_features['INDUS'])
loglstat = np.log(boston_features['LSTAT'])
logptratio = np.log(boston_features['PTRATIO'])

# Min-Max scaling
boston_features['B'] = (b-min(b))/(max(b)-min(b))
boston_features['CRIM'] = (logcrim-min(logcrim))/(max(logcrim)-min(logcrim))
boston_features['DIS'] = (logdis-min(logdis))/(max(logdis)-min(logdis))

# Standardization
boston_features['AGE'] = (age-np.mean(age))/np.sqrt(np.var(age))
boston_features['INDUS'] = (logindus-np.mean(logindus))/np.sqrt(np.var(logindus))
boston_features['LSTAT'] = (loglstat-np.mean(loglstat))/np.sqrt(np.var(loglstat))
boston_features['PTRATIO'] = (logptratio-np.mean(logptratio))/(np.sqrt(np.var(logptratio)))

In [50]:
boston_features.head()

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0  0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1  0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2  0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3  0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4  0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   

      LSTAT  RAD_(6, 24]  TAX_(270, 360]  TAX_(360, 712]  
0 -1.275260            0               1               0  
1 -0.263711            0               0               0  
2 -1.627858            0               0               0  
3 -2.153192            0               0               0  
4 -1.162114            0               0               0

In [51]:
X=boston_features
X

CRIM     INDUS  CHAS     RM       AGE       DIS   PTRATIO         B  \
0    0.000000 -1.704344   0.0  6.575 -0.120013  0.542096 -1.443977  1.000000   
1    0.153211 -0.263239   0.0  6.421  0.367166  0.623954 -0.230278  1.000000   
2    0.153134 -0.263239   0.0  7.185 -0.265812  0.623954 -0.230278  0.989737   
3    0.171005 -1.778965   0.0  6.998 -0.809889  0.707895  0.165279  0.994276   
4    0.250315 -1.778965   0.0  7.147 -0.511180  0.707895  0.165279  1.000000   
..        ...       ...   ...    ...       ...       ...       ...       ...   
501  0.240099  0.410792   0.0  6.593  0.018673  0.331081  1.095518  0.987619   
502  0.206118  0.410792   0.0  6.120  0.288933  0.297277  1.095518  1.000000   
503  0.236926  0.410792   0.0  6.976  0.797449  0.274575  1.095518  1.000000   
504  0.298671  0.410792   0.0  6.794  0.736996  0.315551  1.095518  0.991301   
505  0.210954  0.410792   0.0  6.030  0.434732  0.335545  1.095518  1.000000   

        LSTAT  RAD_(6, 24]  TAX_(270, 360]  TAX_(360, 712]  
0   -1.275260            0               1               0  
1   -0.263711            0               0               0  
2   -1.627858            0               0               0  
3   -2.153192            0               0               0  
4   -1.162114            0               0               0  
..        ...          ...             ...             ...  
501 -0.169811            0               1               0  
502 -0.274682            0               1               0  
503 -1.067939            0               1               0  
504 -0.836660            0               1               0  
505 -0.510809            0               1               0  

[506 rows x 12 columns]

In [52]:
y=pd.DataFrame(boston.target,columns=['price'])
y

price
0     24.0
1     21.6
2     34.7
3     33.4
4     36.2
..     ...
501   22.4
502   20.6
503   23.9
504   22.0
505   11.9

[506 rows x 1 columns]

## Run a linear model in statsmodels

In [53]:
# first method:  using the ols() function from statsmodels.api directly.
# No need to code a formula but the predictors dataframe needs a constant column 

In [62]:
# Your code here
import statsmodels.api as sm

X_int = sm.add_constant(X)
model1= sm.OLS(y,X_int).fit()
model1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  price   R-squared:                       0.779
Model:                            OLS   Adj. R-squared:                  0.774
Method:                 Least Squares   F-statistic:                     144.9
Date:                Thu, 26 Mar 2020   Prob (F-statistic):          5.08e-153
Time:                        16:38:41   Log-Likelihood:                -1458.2
No. Observations:                 506   AIC:                             2942.
Df Residuals:                     493   BIC:                             2997.
Df Model:                          12                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const              8.6442      3.189      2.711      0.007       2.379      14.910
CRIM              -1.9538      2.115     -0.924      0.356      -6.110       2.202
INDUS             -0.8046      0.362     -2.220      0.027      -1.517      -0.093
CHAS               2.5959      0.796      3.260      0.001       1.032       4.160
RM                 2.6466      0.408      6.488      0.000       1.845       3.448
AGE                0.0794      0.352      0.226      0.821      -0.612       0.770
DIS              -10.0962      1.856     -5.439      0.000     -13.743      -6.449
PTRATIO           -1.4867      0.241     -6.160      0.000      -1.961      -1.013
B                  3.8412      0.986      3.897      0.000       1.905       5.778
LSTAT             -5.6288      0.354    -15.912      0.000      -6.324      -4.934
RAD_(6, 24]        1.3380      0.672      1.990      0.047       0.017       2.659
TAX_(270, 360]    -1.2598      0.600     -2.100      0.036      -2.438      -0.081
TAX_(360, 712]    -2.1461      0.704     -3.047      0.002      -3.530      -0.762
==============================================================================
Omnibus:                      106.730   Durbin-Watson:                   1.093
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              432.101
Skew:                           0.891   Prob(JB):                     1.48e-94
Kurtosis:                       7.162   Cond. No.                         117.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [66]:
xtest1 = np.array(X_int.iloc[95])
# xtest1 = np.array(X_int.iloc[95]).reshape(1,-1)

In [67]:
xtest1

array([ 1.        ,  0.30993552, -1.41575367,  0.        ,  6.625     ,
       -0.38316165,  0.47588496, -0.14067514,  0.90186091, -0.79352068,
        0.        ,  1.        ,  0.        ])

In [69]:
model1.predict(xtest1)

array([28.756339])

In [ ]:
#second method:  importing ols. Need to code a formula.
It does not work because of the formula ????


In [19]:
import statsmodels.api as sm
from statsmodels.formula.api import ols
# X.columns :
# ['CRIM', 'INDUS', 'CHAS', 'RM', 'AGE', 'DIS', 'PTRATIO', 'B', 'LSTAT',
#        'RAD_(6, 24]', 'TAX_(270, 360]', 'TAX_(360, 712]']

outcome = 'price'
predictors = X
pred_sum = '+'.join(X.columns)
formula = outcome + '~' + pred_sum
print(formula)

df = pd.concat([y,X])
model = ols(formula=formula, data=df).fit()
model.summary()

price~CRIM+INDUS+CHAS+RM+AGE+DIS+PTRATIO+B+LSTAT+RAD_(6, 24]+TAX_(270, 360]+TAX_(360, 712]


C:\Users\poari\Anaconda3\envs\learn-env\lib\site-packages\ipykernel_launcher.py:13: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  del sys.path[0]


SyntaxError: invalid syntax (<unknown>, line 1)

In [21]:
formula = 'price~CRIM+INDUS+CHAS+RM+AGE+DIS+PTRATIO+B+LSTAT'
df = pd.concat([y,X])
model = ols(formula=formula, data=df).fit()
model.summary()

C:\Users\poari\Anaconda3\envs\learn-env\lib\site-packages\ipykernel_launcher.py:2: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  


ValueError: zero-size array to reduction operation maximum which has no identity

## Run the same model in scikit-learn

In [24]:
# Your code here - Check that the coefficients and intercept are the same as those from Statsmodels
from sklearn.linear_model import LinearRegression

In [25]:
linreg = LinearRegression()
X= boston_features
y=pd.DataFrame(boston['target'], columns = ['price'])


model = linreg.fit(X,y)

In [26]:
model.coef_

array([[ -1.95380233,  -0.80457549,   2.59586776,   2.64657111,
          0.07939727, -10.09618465,  -1.48666599,   3.8412139 ,
         -5.62879369,   1.33796317,  -1.25977612,  -2.14606188]])

In [27]:
model.intercept_

array([8.64415614])

In [28]:
model.rank_

12

In [47]:

xtest = np.array(X.iloc[95]).reshape(1,-1)
print( model.predict(xtest) , y.iloc[95])

[[28.756339]] price    28.4
Name: 95, dtype: float64


In [55]:
x_test

NameError: name 'x_test' is not defined

In [31]:
y

price
0     24.0
1     21.6
2     34.7
3     33.4
4     36.2
..     ...
501   22.4
502   20.6
503   23.9
504   22.0
505   11.9

[506 rows x 1 columns]

## Interpret the coefficients for PTRATIO, PTRATIO, LSTAT

- CRIM: per capita crime rate by town
- INDUS: proportion of non-retail business acres per town
- CHAS: Charles River dummy variable (= 1 if tract bounds river; 0 otherwise)
- RM: average number of rooms per dwelling
- AGE: proportion of owner-occupied units built prior to 1940
- DIS: weighted distances to five Boston employment centres
- RAD: index of accessibility to radial highways
- TAX: full-value property-tax rate per $10,000
- PTRATIO: pupil-teacher ratio by town
- B: 1000(Bk - 0.63)^2 where Bk is the proportion of African American individuals by town
- LSTAT: % lower status of the population

## Predict the house price given the following characteristics (before manipulation!!)

Make sure to transform your variables as needed!

- CRIM: 0.15
- INDUS: 6.07
- CHAS: 1        
- RM:  6.1
- AGE: 33.2
- DIS: 7.6
- PTRATIO: 17
- B: 383
- LSTAT: 10.87
- RAD: 8
- TAX: 284

## Summary
Congratulations! You pre-processed the Boston Housing data using scaling and standardization. You also fitted your first multiple linear regression model on the Boston Housing data using statsmodels and scikit-learn!